external/storm-hbase/src/main/java/org/apache/storm/hbase/state/HBaseKeyValueStateIterator.java

```diff
@@ -0,0 +1,153 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.hbase.state;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.google.common.primitives.UnsignedBytes;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.storm.hbase.common.HBaseClient;
+import org.apache.storm.state.BaseBinaryStateIterator;
+import org.apache.storm.state.DefaultStateEncoder;
+import org.apache.storm.state.Serializer;
+
+import org.apache.storm.state.StateEncoder;
+
+import static org.apache.storm.hbase.state.HBaseKeyValueState.STATE_QUALIFIER;
+
+/**
+ * An iterator over {@link HBaseKeyValueState}.
+ */
+public class HBaseKeyValueStateIterator<K, V> extends BaseBinaryStateIterator<K, V> {
+
+    private final byte[] keyNamespace;
+    private byte[] cursorKey;
+    private final byte[] endScanKey;
+    private final byte[] columnFamily;
+    private final HBaseClient hBaseClient;
+    private final int chunkSize;
+    private final StateEncoder<K, V, byte[], byte[]> encoder;
+
+    private Iterator<Map.Entry<byte[], byte[]>> cachedResultIterator;
+
+    /**
+     * Constructor.
+     *
+     * @param namespace The namespace of State
+     * @param columnFamily The column family of state
+     * @param hBaseClient The instance of HBaseClient
+     * @param pendingPrepareIterator The iterator of pendingPrepare
+     * @param pendingCommitIterator The iterator of pendingCommit
+     * @param chunkSize The size of chunk to get entries from HBase
+     * @param keySerializer The serializer of key
+     * @param valueSerializer The serializer of value
+     */
+    public HBaseKeyValueStateIterator(String namespace, byte[] columnFamily, HBaseClient hBaseClient,
+                                      Iterator<Map.Entry<byte[], byte[]>> pendingPrepareIterator,
+                                      Iterator<Map.Entry<byte[], byte[]>> pendingCommitIterator,
+                                      int chunkSize, Serializer<K> keySerializer,
+                                      Serializer<V> valueSerializer) {
+        super(pendingPrepareIterator, pendingCommitIterator);
+        this.columnFamily = columnFamily;
+        this.keyNamespace = (namespace + "$key:").getBytes();
+        this.cursorKey = (namespace + "$key:").getBytes();
+        this.endScanKey = advanceRow(this.cursorKey);
```

Comment:
Is this the final end key or the end key for the first chunk ? Can you add some clarifying comment here?